ts_src/SilentPaymentTransaction.ts

```diff
@@ -0,0 +1,374 @@
+  foundPayments = foundPayments.concat(
+    performScan(baseSpendPubkey, S, outputsToCheck, null),
+  );
+
+  // Then, scan for each labeled address
+  for (const m of labelNonces) {
+    const { L, Bm } = generateLabelAndAddress(
+      receiverScanPrivkey,
+      baseSpendPubkey,
+      m,
+    );
+
+    const labeledResults = performScan(Bm, S, outputsToCheck, L);
+
+    // Add the label nonce to any found payments for identification
+    labeledResults.forEach(result => {
+      foundPayments.push({ ...result, labelNonce: m });
+    });
+  }
+
+  return foundPayments;
+}
+
+/**
+ * The core scanning logic, performed for a specific spend public key (B_spend).
+ * @param receiverSpendPubkey - G or B_m
+ * @param S
+ * @param outputsToCheck - array of hex xOnly encoded outputs to check
+ * @param labelScalar
+ = */
+function performScan(
+  receiverSpendPubkey: Uint8Array,
+  S: Uint8Array,
+  outputsToCheck: Set<string>,
+  labelScalar: Uint8Array | null, // L (or null for base)
+): SilentOutput[] {
+  const found: SilentOutput[] = [];
+
+  for (let k = 0; k < outputsToCheck.size; k++) {
+    const derivedOutput = deriveSilentOutput(S, receiverSpendPubkey, k);
+    if (!derivedOutput.pub_key) break;
+    const xonlyHex = toHex(derivedOutput.pub_key).toLowerCase();
+
+    if (outputsToCheck.size === 0 || outputsToCheck.has(xonlyHex)) {
+      // priv_key_tweak returned by L + t_k (mod n) for labeled, or t_k for unlabeled
+      let spendTweak = derivedOutput.tweak_key;
+      if (labelScalar != null && !isZero32(labelScalar)) {
+        const sum: Uint8Array | null = ecc.privateAdd(
+          labelScalar,
+          derivedOutput.tweak_key,
+        );
+        if (!sum) throw new Error('privateAdd(label, t_k) failed');
+        spendTweak = sum;
+      }
+      found.push({
+        ...derivedOutput,
+        priv_key_tweak: spendTweak,
+        labelScalar,
+      });
+    }
+  }
+  return found;
+}
+
+/** A Previous Output object, containing the scriptPubKey and value needed to scan an input. */
+export interface Prevout {
+  txid: string; // The transaction id of the previous output.
+  vout: number; // The output index of the previous output.
+  script: Uint8Array; // The scriptPubKey of the previous output.
+  value: number; // The value of the previous output.
+}
+
+/** The receiver's private keys needed for scanning. */
+interface ReceiverKeys {
+  scanPrivKey: Uint8Array;
+  spendPrivKey: Uint8Array;
+}
+
+/**
+ * The result of a successful scan, providing the data needed to spend the output.
+ * Extends the base SilentOutput from p2sp.ts to include transaction-specific UTXO info.
+ */
+export interface SilentOutput extends P2SPSilentOutput {
+  utxo: {
+    txid: string;
+    vout: number;
+    value: number;
+  };
+}
+
+/**
+ * Extends the bitcoinjs-lib Transaction class to provide methods
+ * for creating and scanning for BIP-352 Silent Payments.
+ */
+export class SilentPaymentTransaction extends bitcoin.Transaction {
```

Comment:
Hopefully a class like this makes more sense @junderw 
I do think some functions should change to make it easier and more clear for the developer